### PR TITLE
51 vimba python emlyndev

### PR DESCRIPTION
**Pull request by emlynjdavies**

### CODE REVIEW

File: pysilcam/background.py [hunk 1]
```diff
@@ -6,220 +6,262 @@
...
+                # plt.show()
+                # input()
+                # filename = os.path.join("pyvimba_test", timestamp.strftime('D%Y%m%dT%H%M%S.%f.silc'))
+                # with open(filename, 'wb') as fh:
+                #     np.save(fh, imc, allow_pickle=False)
+                #     fh.flush()
+                #     os.fsync(fh.fileno())
+                if self.proc_image_queue is not None:
+                    logger.debug('Adding image to processing queue: ' + str(timestamp))
+                    addToQueue(self.real_time_stats, self.proc_image_queue, image_number, timestamp, imc)  # the tuple (i, timestamp, imc) is added to the inputQueue
```
Comment (tankefugl) [inline pysilcam/background.py:266]: E501: line too long (165 > 130 characters)

File: pysilcam/background.py [hunk 2]
```diff
@@ -6,220 +6,262 @@
...
+                b = imc[:, :, 2]
+                s = np.std([r, g, b])
+                # ignore bad images
+                if s <= self.bad_lighting_limit:
+                    self.bgstack = bgstack_new
+                    self.imbg = imbg_new
+
+                    if self.proc_image_queue is not None:
+                        logger.debug('Adding image to processing queue: ' + str(timestamp))
+                        addToQueue(self.real_time_stats, self.proc_image_queue, 0, timestamp, imc)  # the tuple (i, timestamp, imc) is added to the inputQueue
```
Comment (tankefugl) [inline pysilcam/background.py:249]: E501: line too long (158 > 130 characters)

File: pysilcam/background.py [hunk 3]
```diff
@@ -6,220 +6,262 @@
...
 '''
-import numpy as np
+import time
 import logging
+import multiprocessing
+import numpy as np
+from pysilcam.acquisition import addToQueue
+
+import os
+import matplotlib.pyplot as plt
```
Comment (tankefugl) [inline pysilcam/background.py:16]: F401: 'matplotlib.pyplot as plt' imported but unused

File: pysilcam/background.py [hunk 4]
```diff
@@ -6,220 +6,262 @@
...
 acquire() must produce a float64 np array
 '''
-import numpy as np
+import time
 import logging
+import multiprocessing
+import numpy as np
+from pysilcam.acquisition import addToQueue
+
+import os
```
Comment (tankefugl) [inline pysilcam/background.py:15]: F401: 'os' imported but unused

File: pysilcam/acquisition.py [hunk 5]
```diff
@@ -138,122 +257,222 @@ def get_generator_disc(self, datapath=None, writeToDisk=False, camera_config_fil
...
+
+
+class MyManager(BaseManager):
+    '''
+    Customized manager class used to register LifoQueues
+    '''
+    pass
+
+
+MyManager.register('LifoQueue', LifoQueue)
```
Comment (tankefugl) [inline pysilcam/acquisition.py:478]: W292: no newline at end of file

File: pysilcam/acquisition.py [hunk 6]
```diff
@@ -138,122 +257,222 @@ def get_generator_disc(self, datapath=None, writeToDisk=False, camera_config_fil
...
+        except:
+            pass
+    else:
+        logger.debug('addToQueue not-realtime')
+        while True:
+            try:
+                inputQueue.put((i, timestamp, imc), True, 0.5)
+                logger.debug(('addToQueue not-realtime:', timestamp))
+                break
+            except:
```
Comment (tankefugl) [inline pysilcam/acquisition.py:416]: E722: do not use bare 'except'

File: pysilcam/acquisition.py [hunk 7]
```diff
@@ -138,122 +257,222 @@ def get_generator_disc(self, datapath=None, writeToDisk=False, camera_config_fil
...
+                                  initilised using defineQueues()
+        i            (int)      : index of the image acquired
+        timestamp    (timestamp): timestamp of the acquired image
+        imc          (uint8)    : corrected image
+    '''
+    if realtime:
+        logger.debug('addToQueue realtime')
+        try:
+            inputQueue.put_nowait((i, timestamp, imc))
+        except:
```
Comment (tankefugl) [inline pysilcam/acquisition.py:407]: E722: do not use bare 'except'

File: pysilcam/acquisition.py [hunk 8]
```diff
@@ -138,122 +257,222 @@ def get_generator_disc(self, datapath=None, writeToDisk=False, camera_config_fil
...
 
-        output = img.copy()
-        
-        return timestamp, output
+        '''
+        print('acquisition    _acquire_frame')
+        if frame.get_status() == FrameStatus.Complete:
+            self.timestamp = pd.Timestamp.now()
+            self.image = frame.as_numpy_ndarray()
+            cam.queue_frame(frame)
```
Comment (tankefugl) [inline pysilcam/acquisition.py:373]: F821: undefined name 'cam'

File: pysilcam/acquisition.py [hunk 9]
```diff
@@ -138,122 +257,222 @@ def get_generator_disc(self, datapath=None, writeToDisk=False, camera_config_fil
...
-        frame0.waitFrameCapture()
-
-        #Copy frame data to numpy array (Bayer format)
-        #bayer_img = np.ndarray(buffer = frame0.getBufferByteData(),
-        #                       dtype = np.uint8,
-        #                       shape = (frame0.height, frame0.width, 3))
-        img = np.ndarray(buffer = frame0.getBufferByteData(),
-                        dtype = np.uint8,
-                        shape = (frame0.height, frame0.width, 3))
+            except:
```
Comment (tankefugl) [inline pysilcam/acquisition.py:357]: E722: do not use bare 'except'

File: pysilcam/acquisition.py [hunk 10]
```diff
@@ -138,122 +257,222 @@ def get_generator_disc(self, datapath=None, writeToDisk=False, camera_config_fil
...
+            while (self.gui.qsize() > 0):
+                try:
+                    self.gui.get_nowait()
+                    time.sleep(0.001)
+                except:
+                    continue
+
+            rtdict = dict()
+            rtdict = {'dias': 0,
+                        'vd_oil': 0,
```
Comment (tankefugl) [inline pysilcam/acquisition.py:322]: E127: continuation line over-indented for visual indent

File: pysilcam/acquisition.py [hunk 11]
```diff
@@ -138,122 +257,222 @@ def get_generator_disc(self, datapath=None, writeToDisk=False, camera_config_fil
...
+
+            camera.queue_frame(frame)  # ask the camera for the next frame, which would evtentually call image_handler again
+
+    def gui_update(self, timestamp, img):
+        if self.gui is not None:
+            while (self.gui.qsize() > 0):
+                try:
+                    self.gui.get_nowait()
+                    time.sleep(0.001)
+                except:
```
Comment (tankefugl) [inline pysilcam/acquisition.py:317]: E722: do not use bare 'except'

File: pysilcam/acquisition.py [hunk 12]
```diff
@@ -138,122 +257,222 @@ def get_generator_disc(self, datapath=None, writeToDisk=False, camera_config_fil
...
+
+                # gui data handling
+                self.gui_update(timestamp, img)
+
+                if self.raw_image_queue is not None:
+                    # Prev put_nowait() !!!!!
+                    logger.debug(f"timestamp+img adding to stack: {datetime.now()}")
+                    try:
+                        self.raw_image_queue.put_nowait((-1, timestamp, img))
+                    except:
```
Comment (tankefugl) [inline pysilcam/acquisition.py:304]: E722: do not use bare 'except'

File: pysilcam/acquisition.py [hunk 13]
```diff
@@ -138,122 +257,222 @@ def get_generator_disc(self, datapath=None, writeToDisk=False, camera_config_fil
...
 
-        Yields:
-            timestamp: timestamp of the acquired image.
-            img: acquired image.
+                timestamp = pd.Timestamp.now()
+                filename = os.path.join(self.datapath, timestamp.strftime('D%Y%m%dT%H%M%S.%f.silc'))
+
+                if self.writeToDisk:
+                    with open(filename, 'wb') as fh:
+                        np.save(fh, img, allow_pickle = False)
```
Comment (tankefugl) [inline pysilcam/acquisition.py:290]: E251: unexpected spaces around keyword / parameter equals

Comment (tankefugl) [inline pysilcam/acquisition.py:290]: E251: unexpected spaces around keyword / parameter equals

File: pysilcam/acquisition.py [hunk 14]
```diff
@@ -97,30 +106,140 @@ def print_camera_config(camera):
...
+            self.gui_update(timestamp, im_raw)
+
+        logger.info('end of file list')
+        while True:
+            try:
+                logger.debug("Putting None into raw_image_queue")
+                self.raw_image_queue.put(None, True, 0.5)
+                logger.debug("None put into raw_image_queue")
+                break
+            except:
```
Comment (tankefugl) [inline pysilcam/acquisition.py:233]: E722: do not use bare 'except'

File: pysilcam/acquisition.py [hunk 15]
```diff
@@ -97,30 +106,140 @@ def print_camera_config(camera):
...
+
         if USE_PYMBA:
-            self.pymba = pymba
-            self.pymba.get_time_stamp = lambda x: pd.Timestamp.now()
-            logger.info('Pymba imported')
-            self.get_generator = self.get_generator_camera
+            self.vimba = Vimba
+            logger.info('Vimba imported')
+
+            if datapath != None:
```
Comment (tankefugl) [inline pysilcam/acquisition.py:150]: E711: comparison to None should be 'if cond is not None:'

File: pysilcam/acquisition.py [hunk 16]
```diff
@@ -97,30 +106,140 @@ def print_camera_config(camera):
...
+
+
 class Acquire():
     '''
     Class used to acquire images from camera or disc
     '''
-    def __init__(self, USE_PYMBA=False, FAKE_PYMBA_OFFSET=0):
+    def __init__(self, USE_PYMBA=False, datapath=None, writeToDisk=False,
+                 FAKE_PYMBA_OFFSET=0, gui=None, raw_image_queue=None):
+
```
Comment (tankefugl) [inline pysilcam/acquisition.py:141]: W293: blank line contains whitespace

File: pysilcam/acquisition.py [hunk 17]
```diff
@@ -97,30 +106,140 @@ def print_camera_config(camera):
     config_info = '\n'.join(['{0}: {1}'.format(a, camera.getattr(a))
                              for a, b in config_info_map])
 
-    logger.info(config_info) # TODO: Should this be printed?
+    logger.info(config_info)  # TODO: Should this be printed?
     logger.debug(config_info)
 
 
+def silcam_load(filename):
+    #Load the raw image from disc depending on filetype
```
Comment (tankefugl) [inline pysilcam/acquisition.py:114]: E265: block comment should start with '# '

File: pysilcam/acquisition.py [hunk 18]
```diff
@@ -4,77 +4,86 @@
...
+        config = load_config(config_file)
+
+        # All settings from the pysilcam config ini files can now be applied.
+        for k, v in config['Camera'].items():
+            print(k, v)
+            logger.info('{0} = {1}'.format(k, v))
+            # try to write settings to the camera
+            try:
+                getattr(camera, k).set(v)
+            except AttributeError:  # if there is an element of the camera config that is not compatible, then continue to the next
```
Comment (tankefugl) [inline pysilcam/acquisition.py:83]: E501: line too long (131 > 130 characters)

File: pysilcam/__main__.py [hunk 19]
```diff
@@ -156,45 +147,66 @@ def silcam_acquire(datapath, config_filename, writeToDisk=True, gui=None):
...
+    acq.stream_images(config_file=config_filename)
+
+    backgrounder_process.join()  # shut down subprocesses after stopping (needs checking how to do this)
 
 
 # the standard processing method under active development
-def silcam_process(config_filename, datapath, multiProcess=True, realtime=False, discWrite=False, nbImages=None,
-                   gui=None,
-                   overwriteSTATS=True):
+def silcam_process(config_filename, datapath, multiProcess=True, realtime=False,
```
Comment (tankefugl) [inline pysilcam/__main__.py:208]: C901: 'silcam_process' is too complex (12)

**Review by tankefugl — CHANGES_REQUESTED** [review]
[Lintly](https://github.com/grantmcconnaughey/Lintly) has detected code quality issues in this pull request.

File: pysilcam/background.py (hunk 1 above)
Comment (tankefugl) [inline pysilcam/background.py:266]: E501: line too long (165 > 130 characters)

File: pysilcam/background.py (hunk 2 above)
Comment (tankefugl) [inline pysilcam/background.py:249]: E501: line too long (158 > 130 characters)

File: pysilcam/background.py (hunk 3 above)
Comment (tankefugl) [inline pysilcam/background.py:16]: F401: 'matplotlib.pyplot as plt' imported but unused

File: pysilcam/background.py (hunk 4 above)
Comment (tankefugl) [inline pysilcam/background.py:15]: F401: 'os' imported but unused

File: pysilcam/acquisition.py (hunk 5 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:478]: W292: no newline at end of file

File: pysilcam/acquisition.py (hunk 6 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:416]: E722: do not use bare 'except'

File: pysilcam/acquisition.py (hunk 7 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:407]: E722: do not use bare 'except'

File: pysilcam/acquisition.py (hunk 8 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:373]: F821: undefined name 'cam'

File: pysilcam/acquisition.py (hunk 9 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:357]: E722: do not use bare 'except'

File: pysilcam/acquisition.py (hunk 10 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:322]: E127: continuation line over-indented for visual indent

File: pysilcam/acquisition.py (hunk 11 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:317]: E722: do not use bare 'except'

File: pysilcam/acquisition.py (hunk 12 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:304]: E722: do not use bare 'except'

File: pysilcam/acquisition.py (hunk 13 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:290]: E251: unexpected spaces around keyword / parameter equals

Comment (tankefugl) [inline pysilcam/acquisition.py:290]: E251: unexpected spaces around keyword / parameter equals

File: pysilcam/acquisition.py (hunk 14 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:233]: E722: do not use bare 'except'

File: pysilcam/acquisition.py (hunk 15 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:150]: E711: comparison to None should be 'if cond is not None:'

File: pysilcam/acquisition.py (hunk 16 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:141]: W293: blank line contains whitespace

File: pysilcam/acquisition.py (hunk 17 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:114]: E265: block comment should start with '# '

File: pysilcam/acquisition.py (hunk 18 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:83]: E501: line too long (131 > 130 characters)

File: pysilcam/__main__.py (hunk 19 above)
Comment (tankefugl) [inline pysilcam/__main__.py:208]: C901: 'silcam_process' is too complex (12)

**Review by tankefugl — CHANGES_REQUESTED** [review]
[Lintly](https://github.com/grantmcconnaughey/Lintly) has detected code quality issues in this pull request.

File: pysilcam/background.py [hunk 20]
```diff
@@ -6,220 +6,262 @@
...
+                # plt.show()
+                # input()
+                # filename = os.path.join("pyvimba_test", timestamp.strftime('D%Y%m%dT%H%M%S.%f.silc'))
+                # with open(filename, 'wb') as fh:
+                #     np.save(fh, imc, allow_pickle=False)
+                #     fh.flush()
+                #     os.fsync(fh.fileno())
+                if self.proc_image_queue is not None:
+                    logger.debug('Adding image to processing queue: ' + str(timestamp))
+                    addToQueue(self.real_time_stats, self.proc_image_queue, image_number, timestamp, imc)  # the tuple (i, timestamp, imc) is added to the inputQueue
```
Comment (tankefugl) [inline pysilcam/background.py:266]: E501: line too long (165 > 130 characters)

File: pysilcam/background.py [hunk 21]
```diff
@@ -6,220 +6,262 @@
...
+                b = imc[:, :, 2]
+                s = np.std([r, g, b])
+                # ignore bad images
+                if s <= self.bad_lighting_limit:
+                    self.bgstack = bgstack_new
+                    self.imbg = imbg_new
+
+                    if self.proc_image_queue is not None:
+                        logger.debug('Adding image to processing queue: ' + str(timestamp))
+                        addToQueue(self.real_time_stats, self.proc_image_queue, 0, timestamp, imc)  # the tuple (i, timestamp, imc) is added to the inputQueue
```
Comment (tankefugl) [inline pysilcam/background.py:249]: E501: line too long (158 > 130 characters)

File: pysilcam/background.py (hunk 3 above)
Comment (tankefugl) [inline pysilcam/background.py:16]: F401: 'matplotlib.pyplot as plt' imported but unused

File: pysilcam/background.py (hunk 4 above)
Comment (tankefugl) [inline pysilcam/background.py:15]: F401: 'os' imported but unused

File: pysilcam/acquisition.py (hunk 5 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:478]: W292: no newline at end of file

File: pysilcam/acquisition.py (hunk 6 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:416]: E722: do not use bare 'except'

File: pysilcam/acquisition.py (hunk 7 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:407]: E722: do not use bare 'except'

File: pysilcam/acquisition.py (hunk 8 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:373]: F821: undefined name 'cam'

File: pysilcam/acquisition.py (hunk 9 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:357]: E722: do not use bare 'except'

File: pysilcam/acquisition.py (hunk 10 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:322]: E127: continuation line over-indented for visual indent

File: pysilcam/acquisition.py (hunk 11 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:317]: E722: do not use bare 'except'

File: pysilcam/acquisition.py (hunk 12 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:304]: E722: do not use bare 'except'

File: pysilcam/acquisition.py (hunk 13 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:290]: E251: unexpected spaces around keyword / parameter equals

Comment (tankefugl) [inline pysilcam/acquisition.py:290]: E251: unexpected spaces around keyword / parameter equals

File: pysilcam/acquisition.py (hunk 14 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:233]: E722: do not use bare 'except'

File: pysilcam/acquisition.py (hunk 15 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:150]: E711: comparison to None should be 'if cond is not None:'

File: pysilcam/acquisition.py (hunk 16 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:141]: W293: blank line contains whitespace

File: pysilcam/acquisition.py (hunk 17 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:114]: E265: block comment should start with '# '

File: pysilcam/acquisition.py (hunk 18 above)
Comment (tankefugl) [inline pysilcam/acquisition.py:83]: E501: line too long (131 > 130 characters)

File: pysilcam/__main__.py [hunk 22]
```diff
@@ -156,45 +147,66 @@ def silcam_acquire(datapath, config_filename, writeToDisk=True, gui=None):
...
```
Comment (tankefugl) [inline pysilcam/__main__.py:208]: C901: 'silcam_process' is too complex (12)

**Review by tankefugl — CHANGES_REQUESTED** [review]
[Lintly](https://github.com/grantmcconnaughey/Lintly) has detected code quality issues in this pull request.

**Comment by emlynjdavies** [comment]
only opened for testing

**Pull request closed without merging**